jme3-core/src/main/resources/Common/ShaderLib/module/pbrlighting/PBRLightingUtils.glsllib

```diff
@@ -0,0 +1,592 @@
+        #endif
+        
+        // computeSpotFalloff
+        if(light.type>1.){
+            vec3 spotdir = normalize(light.spotDirection);
+            float curAngleCos = dot(-L, spotdir);    
+            float innerAngleCos = floor(light.spotAngleCos) * 0.001;
+            float outerAngleCos = fract(light.spotAngleCos);
+            float innerMinusOuter = innerAngleCos - outerAngleCos;
+            float falloff = clamp((curAngleCos - outerAngleCos) / innerMinusOuter, 0.0, 1.0);
+            #ifdef SRGB
+                // Use quadratic falloff (notice the ^4)
+                falloff = pow(clamp((curAngleCos - outerAngleCos) / innerMinusOuter, 0.0, 1.0), 4.0);
+            #endif
+            light.fallOff*=falloff;
+        }
+
+
+        vec3 h=normalize(L+viewDir);
+        light.dir=L;
+        light.NdotL = max(dot(worldNormal, L), 0.0);
+        light.NdotH = max(dot(worldNormal, h), 0.0);
+        light.LdotH = max(dot(L, h), 0.0);
+        light.HdotV = max(dot(viewDir,h), 0.);
+    }
+#endif
+
+
+#ifdef ENABLE_PBRLightingUtils_readPBRSurface
+    PBRSurface PBRLightingUtils_readPBRSurface(    
+        in vec3 wViewDir
+    ){
+        
+        PBRSurface surface;
+        surface.position = wPosition;
+        surface.viewDir = wViewDir;
+        surface.geometryNormal = normalize(wNormal);
+
+        surface.bakedLightContribution = vec3(0);
+        surface.directLightContribution = vec3(0);
+        surface.envLightContribution = vec3(0);
+
+        #if defined(NORMALMAP) || defined(PARALLAXMAP)
+            vec3 tan = normalize(wTangent.xyz);
+            mat3 tbnMat = mat3(tan, wTangent.w * cross( surface.geometryNormal, tan), surface.geometryNormal);
+        #endif
+        
+        vec2 newTexCoord;
+        #if (defined(PARALLAXMAP) || (defined(NORMALMAP_PARALLAX) && defined(NORMALMAP)))
+            vec3 vViewDir =  wViewDir * tbnMat;  
+            #ifdef STEEP_PARALLAX
+                #ifdef NORMALMAP_PARALLAX
+                    //parallax map is stored in the alpha channel of the normal map         
+                    newTexCoord = steepParallaxOffset(m_NormalMap, vViewDir, texCoord, m_ParallaxHeight);
+                #else
+                    //parallax map is a texture
+                    newTexCoord = steepParallaxOffset(m_ParallaxMap, vViewDir, texCoord, m_ParallaxHeight);         
+                #endif
+            #else
+                #ifdef NORMALMAP_PARALLAX
+                    //parallax map is stored in the alpha channel of the normal map         
+                    newTexCoord = classicParallaxOffset(m_NormalMap, vViewDir, texCoord, m_ParallaxHeight);
+                #else
+                //parallax map is a texture
+                newTexCoord = classicParallaxOffset(m_ParallaxMap, vViewDir, texCoord, m_ParallaxHeight);
+            #endif
+        #endif
+        #else
+        newTexCoord = texCoord;    
+        #endif
+        
+        #ifdef BASECOLORMAP

```

Comment:
Yes that would increase the modularity, although I didn't do this at first because I figured iit would be unlikely that a shader dev would need to write their shader to do an albedo texture read without having code to do the rest of the standard pbr texture reads, since typically you'd just leave the un-wanted textures un-defined to achieve the same thing.

But from a code standpoint, it would certainly look more organized if each texture read were in its own function, so I will update the .glsllibs to put each texture read in its own function.